Add Index.validate_dataarray_coord

benbovy opened the pull request:
<!-- Feel free to remove check-list items aren't relevant to your change -->

- [x] Closes #10116
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [x] New functions/methods are listed in `api.rst`

Same goal than #10116 using the alternative approach suggested in https://github.com/pydata/xarray/pull/10116#issuecomment-2725004683 where the propagation (validation) of coordinates in a DataArray is delegated to their index (if any).

I find this approach cleaner than #10116. Here is the same [notebook](https://notebooksharing.space/view/9dbdbd9b3fe71c6dc06fbecaa273934487057fdfccb85731837e5262dc58cd6a#displayOptions=) example adapted for this PR.

### Index API alternatives

1. `Index.validate_dataarray_coord(self, name: Hashable, var: Variable, dims: set[Hashable]) -> None`
    - check one index coordinate at a time, thus called multiple times for a multi-coordinate index

2. `Index.validate_dataarray_coords(self, variables: dict[Hashable, Variable], dims: set[Hashable]) -> None`
    - check all coordinates of an index at once
    - perhaps could make things easier for developers of custom Xarray indexes (perhaps not?), but it would make Xarray internals more complicated with possible significant impact on performance (we'd have to rely on `Indexes.group_by_index` while in option 1 we simply need to iterate over `_variables`, `_coords` and/or `_indexes`).

3. `Index.validate_dataarray_coords(self, variables: dict[Hashable, Variable], dims: set[Hashable]) -> Coordinates`
    - Same than option 2 but allow returning a new set of coordinate variables and their indexes.
    - This has much more flexibility (e.g., convert a custom 2D spatial index to a 1D default PandasIndex when the result has only one of the two spatial dimensions), but I find it also complicated and prone to confusion.

Option 1, the simplest and working well with IntervalIndex, is currently implemented in this PR.

cc @shoyer @dcherian

shoyer commented on xarray/core/dataset.py:
```diff
@@ -1212,7 +1212,18 @@ def _construct_dataarray(self, name: Hashable) -> DataArray:
+                        k, self._variables[k], needed_dims
+                    )
+                    coords[k] = self._variables[k]
+                except ValueError:
```
`ValueError` can be raised by many different things going wrong in Python. If we're going to check specifically for an error, we should only catch a specific Exception subclass.

benbovy replied:
Added `CoordinateValidationError` in [5da014e](https://github.com/pydata/xarray/pull/10137/commits/5da014e260576b8c0579b312e0bde42a6461e5eb).

For now it is defined in `xarray/core/indexes.py` to avoid a cyclic import. That's a bit messy but we could maybe move the definition of all Xarray custom error types in a dedicated `xarray.exception` module (in a later PR?).

shoyer commented on xarray/core/dataset.py:
```diff
@@ -1212,7 +1212,18 @@ def _construct_dataarray(self, name: Hashable) -> DataArray:
+                    )
+                    coords[k] = self._variables[k]
+                except ValueError:
+                    # failback to strict DataArray model check (index may be dropped later)
+                    if var_dims <= needed_dims:
+                        coords[k] = self._variables[k]
```
It is a little weird how the fallback logic here doesn't depend on the index, but is called inside the loop.

benbovy replied:
I'm not sure to understand? Maybe you've seen `if k in self._indexes` above instead of `for k in self._indexes` above?

shoyer replied:
oops, indeed, I misread this!

shoyer commented on xarray/core/dataset.py:
```diff
@@ -1212,7 +1212,18 @@ def _construct_dataarray(self, name: Hashable) -> DataArray:
         coords: dict[Hashable, Variable] = {}
         # preserve ordering
         for k in self._variables:
-            if k in self._coord_names and set(self._variables[k].dims) <= needed_dims:
+            var_dims = set(self._variables[k].dims)
+            if k in self._indexes:
```
I am a little concerned from a performance perspective about introducing a double loop over variables/indexes inside DataArray creation.

I need to think a little harder about what the alternative would be, though...

shoyer reviewed (commented):
Some quick thoughts

benbovy commented:
Thanks @shoyer for taking a look!

This is now ready for (another round of) review.

Compared to #10116 this PR still represents a major data model change for DataArray (in the sense that it allows overriding strict enforcement of the DataArray model in certain cases), but I think that the risk is mitigated since it is here done explicitly via `Index.validate_dataarray_coord` and the default behavior is to keep conforming to the DataArray model. Limiting an index to (in)validate its coordinates may also be good to keep things simple and predictable for end-users.

@dcherian regarding your https://github.com/pydata/xarray/pull/10116#issuecomment-2724920394, in the example notebook the DataArray reprs shown without any list of dimensions for the Coordinates section do not strike me much. That said, the `Coordinates` repr is still basic and might benefit from such addition (it doesn't have any html repr BTW).

dcherian commented on xarray/core/indexes.py:
```diff
@@ -195,6 +195,59 @@ def create_variables(
+    ):
+        """Validate an index coordinate variable to include in a DataArray.
+
+        This method is called repeatedly for each coordinate associated with
```
```suggestion
        This method is called repeatedly for each Variable associated with
```

dcherian commented on xarray/core/indexes.py:
```diff
@@ -195,6 +195,59 @@ def create_variables(
+        If the validation passes (i.e., no error raised), the coordinate will be
+        included in the DataArray regardless of its dimensions.
+
+        When a DataArray is constructed from a Dataset (variable access), if the
+        validation fails Xarray will fail back to propagating the coordinate
+        according to the default rules for DataArray (i.e., depending on its
+        dimensions), which may drop this index.
```
```suggestion
        If this method raises when a DataArray is constructed from a Dataset,  
        Xarray will fail back to propagating the coordinate
        according to the default rules for DataArray --- i.e., the dimensions of every 
        coordinate variable must be a subset of DataArray.dims), which may drop this index.
```

dcherian reviewed (approved):
Nice!

dcherian commented:
> That said, the Coordinates repr is still basic and might benefit from such addition (it doesn't have any html repr BTW). 

Yeah I think it could be improved, but certainly not a blocker here.

benbovy commented:
Thanks @dcherian and @shoyer. I added "plan to merge" but I'm open to revisit anything here.

shoyer commented:
Thanks @benbovy for this change, and apologies for the slow review.

My one big question is if `validate_dataarray_coord` could be replaced by a method that returns a boolean rather than potentially raising an error. It's not clear to me what the advantages of raising an error are.

The fall-back logic of checking coordinate dimensions could also potentially be implemented on the base `Index` class, which would allow for customizing it if desired.

benbovy commented:
@shoyer I like your suggestion and find it cleaner.

In [3e55af0](https://github.com/pydata/xarray/pull/10137/commits/3e55af003bbd4b73402a2bd6ae201e7da6972afd) I replaced the `validate_dataarray_coord` with `should_add_coord_in_dataarray` (couldn't find a better name, suggestions welcome!).

I also removed the fall-back in `Dataset._construct_dataarray`. The logic of checking coordinate dimensions is implemented in the `Index` base class (in `Index.should_add_coord_in_dataarray`).

Is that what you had in mind?

dcherian commented:
Yeah this method's a hard one to name. The only thing that comes to mind is `should_propagate_variable_with_dataarray` :/

benbovy commented:
Another possible name: `is_coord_needed_in_dataarray()`

dcherian commented on doc/api-hidden.rst:
```diff
@@ -520,6 +520,7 @@
    Index.stack
    Index.unstack
    Index.create_variables
+   Index.should_add_coord_in_dataarray
```
```suggestion
   Index.should_add_coord_to_array
```

dcherian commented on doc/whats-new.rst:
```diff
@@ -29,6 +29,10 @@ New Features
   includes ``datatree`` support, and removing slashes from dimension names. By
   `Miguel Jimenez-Urias <https://github.com/Mikejmnez>`_.
+- Allow assigning index coordinates with non-array dimension(s) in a :py:class:`DataArray` by overriding
+  :py:meth:`Index.should_add_coord_in_dataarray`. For example, this enables support for CF boundaries coordinate (e.g.,
```
```suggestion
  :py:meth:`Index.should_add_coord_to_array`. For example, this enables support for CF boundaries coordinate (e.g.,
```

dcherian commented on xarray/core/coordinates.py:
```diff
@@ -1154,9 +1146,58 @@ def create_coords_with_default_indexes(
+
+    for k, v in coords.items():
+        if k in indexes:
+            invalid = not indexes[k].should_add_coord_in_dataarray(k, v, dim_set)
```
```suggestion
            invalid = not indexes[k].should_add_coord_to_array(k, v, dim_set)
```

dcherian commented on xarray/core/dataset.py:
```diff
@@ -1159,7 +1159,15 @@ def _construct_dataarray(self, name: Hashable) -> DataArray:
         for k in self._variables:
-            if k in self._coord_names and set(self._variables[k].dims) <= needed_dims:
+            if k in self._indexes:
+                add_coord = self._indexes[k].should_add_coord_in_dataarray(
```
```suggestion
                add_coord = self._indexes[k].should_add_coord_to_aarray(
```

dcherian commented on xarray/core/indexes.py:
```diff
@@ -196,6 +196,50 @@ def create_variables(
         else:
             return {}
 
+    def should_add_coord_in_dataarray(
```
```suggestion
    def should_add_coord_to_array(
```

dcherian commented on xarray/tests/test_dataarray.py:
```diff
@@ -529,6 +537,25 @@ class CustomIndex(Index): ...
 
+    def test_constructor_extra_dim_index_coord(self) -> None:
+        class AnyIndex(Index):
+            def should_add_coord_in_dataarray(self, name, var, dims):
```
```suggestion
            def should_add_coord_to_array(self, name, var, dims):
```

dcherian commented on xarray/tests/test_dataarray.py:
```diff
@@ -1634,6 +1661,27 @@ def test_assign_coords_no_default_index(self) -> None:
 
+    def test_assign_coords_extra_dim_index_coord(self) -> None:
+        class AnyIndex(Index):
+            def should_add_coord_in_dataarray(self, name, var, dims):
```
```suggestion
            def should_add_coord_to_array(self, name, var, dims):
```

dcherian commented on xarray/tests/test_dataset.py:
```diff
@@ -4256,6 +4256,26 @@ def test_getitem_multiple_dtype(self) -> None:
 
+    def test_getitem_extra_dim_index_coord(self) -> None:
+        class AnyIndex(Index):
+            def should_add_coord_in_dataarray(self, name, var, dims):
```
```suggestion
            def should_add_coord_to_array(self, name, var, dims):
```

dcherian commented:
I added suggestions for `should_add_coord_to_array` to save a few characters

shoyer commented on xarray/core/indexes.py:
```diff
@@ -196,6 +196,50 @@ def create_variables(
+        var : Variable
+            Coordinate variable object.
+        dims: tuple
+            Dataarray's dimensions.
```
```suggestion
            This DataArray's dimensions.
```

benbovy replied:
Changed `Dataarray's dimensions` to `Dimensions of the new DataArray object being created`.

shoyer commented on xarray/core/indexes.py:
```diff
@@ -196,6 +196,50 @@ def create_variables(
+        var: Variable,
+        dims: set[Hashable],
+    ) -> bool:
+        """Define whether or not an index coordinate variable should be added in
+        a new DataArray.
+
+        This method is called repeatedly for each Variable associated with
+        this index when creating a new DataArray (via its constructor or from a
+        Dataset) or updating an existing one.
```
Can you define what "associated to this index" means more precise?

benbovy replied:
I added a sentence just after this one.

shoyer commented on xarray/core/indexes.py:
```diff
@@ -195,6 +195,50 @@ def create_variables(
+        Parameters
+        ----------
+        name : Hashable
+            Name of a coordinate variable associated to this index.
```
Oh, you're very right. Yes, there is no `Variable.name` and we should not rely on `IndexVariable.name`.
